README.md
# Many-Sorts
Sorts by way of multiple sorts.
Code written by Alexander Burton.
with some resources from:
https://en.wikipedia.org/wiki/Quicksort


Changelog
Version History:
Date:		||	Time:	||	Version:
04/26/2019	||	05:37PM	||	0.0.0.0		|| This is a sorting class(cont)
designed for Windows 10. It is intended to sort multiple objects of the same type. Multiple sorting algorithms are planned for implementation.

04/26/2019	||	06:55PM	||	0.2.0.0
two functions have been implemented, but yet untested. Quicksort Algorithm is implemented.

04/27/2019	||	06:55PM	||	0.2.1.2
quicksort algorithm is fixed	and selection sort is implemented.

04/26/2019	||	05:37PM	||	0.8.8.8
8 functions have been implemented and fixed in this version.

04/28/2019	||	06:33PM	||	0.8.8.9
Fixed an issue with Quicksort


SourceCode/SortDriver.cpp
/*--------------------------------------------------------------------------||
||	Programmer:			Alexander Burton									||
||	Date Created:		4/26/2019											||
||	Version:			0.8.8.8												||
||	File:				SortDriver.cpp										||
||	Compiler Info:															||
||		Hardware:															||
||			MB:																||
||				Manu:		BaseBoard Manufacturer	ASUSTeK COMPUTER INC.	||
||				Chipset:	BaseBoard Product	P8Z77-V PRO					||
||																			||
||			CPU:															||
||				Chipset:	Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz			||
||				Clock Rate:	3401 Mhz										||
||				Cores:		4												||
||				processors:	8												||
||																			||
||			RAM:															||
||				Manu:				Kingston								||
||				Capacity(indiv):	4GB										||
||				Count:				4										||
||				Capacity(total):	16.0GB									||
||				Type:				240-Pin DDR3 SDRAM DDR3 1333 (PC3 10600)||
||																			||
||		Software:															||
||				OS:			Windows 10 Home Edition							||
||				IDE:		Visual Studio 2019								||
||																			||
||	Functions:																||
||																			||
||		void quicksort(int* list, int low, int high);						||
||		int partition(int* list, int low, int high);						||
||		void selectionSort(int* list,int low, int size);					||
||		void optQuicksort(int* list, int low, int high, int ssSize);		||
||		int optPartition(int* list, int low, int high);						||
||		void optSelectionSort(int* list, int low, int size);				||
||																			||
||		void randomizeArr(int* list, int size);								||
||		void printArr(int* arr, int size);									||
||																			||
||	Variables:																||
||		N/A																	||
||	Notes:																	||
||	Version History:														||
||	Date:		||	Time:	||	Version:	||	Description:				||
||	04/26/2019	||	05:37PM	||	0.0.0.0		|| This is a sorting class(cont)||
||																			||
----------------------------------------------------------------------------*/
#include <iostream>
#include <fstream>
#include <cstdlib>
#include <time.h>
#include "ABSortingClass.h"
//#include "ABFileHandler.h" //being built
const string humanWelcome = "Thank you for using this program!";
const string humanWelcome2 = " Coded by Alexander Burton.\n";
const string makeChoice = "Which sort would you like to choose?";
const string ssChoice = "A) ";
const string ssName = "Selection Sort";
const string qsChoice = "B) ";
const string qsName = "Quick Sort";
const string oqsChoice = "C) ";
const string oqsName = "Optimized Quick Sort";
const string itemSizeString = "How many items would you like sorted?";
const string itemRange = "The number can be between 1 and ";
using namespace std;
using namespace AB_SortingClass;
int main()
{
	bool goodInput = true;
	char choice = ' ';
	const int SSSIZE = 10;
	int itemSize = 0;
	int exitStatus = EXIT_SUCCESS;
	int RandomNumbers[MAX_SIZE];
	cout << humanWelcome << humanWelcome2 << endl;
	cout << makeChoice << endl << ssChoice << ssName << endl << qsChoice <<
		qsName << endl << oqsChoice << oqsName << endl;
	cin >> choice;
	while (cin.fail())
	{
		cout << "Please enter a proper value.\n";
		cin >> choice;
	}
	while (((toupper(choice) < 'A' || toupper(choice) > 'C'))
		&& ((choice < '1') || (choice > '3')))
	{
		cout << "Values are out of range." <<
			" Please choose a number between 1 and 3, or A,B, or C" << endl;
		cin >> choice;
		while (cin.fail())
		{
			cout << "Please enter a proper value.\n";
			cin >> choice;
		}
	}
	cout << itemSizeString << endl << itemRange << MAX_SIZE<<endl;
	cin >> itemSize;
	while (cin.fail())
	{
		cout << "Please enter a proper value.\n";
		cin >> itemSize;
	}
	while ((itemSize > MAX_SIZE) || (itemSize < 1))
	{
		cout << "please enter a value between 1 and " << MAX_SIZE<<endl;
		while (cin.fail())
		{
			cout << "Please enter a proper value.\n";
			cin >> itemSize;
		}
	}
	cout << "Randomizing Array." << endl;
	//fill array sequentially.
	for (int i = 0; i < MAX_SIZE; ++i)
	{
		RandomNumbers[i] = (i + 1);
	}
	cout << "Initializing ";
	choice = toupper(choice);
	switch (choice)
	{
	case '1':
	case 'A':
		cout << ssName << "." << endl;
		selectionSort(RandomNumbers, 0, itemSize);
		break;
	case '2':
	case 'B':
		cout << qsName << "." << endl;
		quicksort(RandomNumbers, 0, itemSize-1);
		break;
	case '3':
	case 'C':
		cout << oqsName << "." << endl;
		optQuicksort(RandomNumbers, 0, itemSize-1, SSSIZE);
		break;
	}
	cout << "Sorting completed." << endl;
	cout << "Printing Array." << endl;
	printArr(RandomNumbers, itemSize);
	cout << endl << "Thank you for using this program." << 
		endl << "-Alexander Burton" << endl;
	cin.ignore();
	cin.get();
	return exitStatus;
}
//--------------------------------End of File---------------------------------

SourceCode/ABSortingClass.h
/*--------------------------------------------------------------------------||
||	Programmer:			Alexander Burton									||
||	Date Created:		4/26/2019											||
||	Version:			0.8.8.8												||
||	File:				ABSortingClass.h									||
||	Compiler Info:															||
||		Hardware:															||
||			MB:																||
||				Manu:		BaseBoard Manufacturer	ASUSTeK COMPUTER INC.	||
||				Chipset:	BaseBoard Product	P8Z77-V PRO					||
||																			||
||			CPU:															||
||				Chipset:	Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz			||
||				Clock Rate:	3401 Mhz										||
||				Cores:		4												||
||				processors:	8												||
||																			||
||			RAM:															||
||				Manu:				Kingston								||
||				Capacity(indiv):	4GB										||
||				Count:				4										||
||				Capacity(total):	16.0GB									||
||				Type:				240-Pin DDR3 SDRAM DDR3 1333 (PC3 10600)||
||																			||
||		Software:															||
||				OS:			Windows 10 Home Edition							||
||				IDE:		Visual Studio 2019								||
||																			||
||	Functions:																||
||																			||
||		void quicksort(int* list, int low, int high);						||
||		int partition(int* list, int low, int high);						||
||		void selectionSort(int* list,int low, int size);					||
||		void optQuicksort(int* list, int low, int high, int ssSize);		||
||		int optPartition(int* list, int low, int high);						||
||		void optSelectionSort(int* list, int low, int size);				||
||																			||
||		void randomizeArr(int* list, int size);								||
||		void printArr(int* arr, int size);									||
||																			||
||	Variables:																||
||		N/A																	||
||	Notes:																	||
||	Version History:														||
||	Date:		||	Time:	||	Version:	||	Description:				||
||	04/26/2019	||	05:37PM	||	0.0.0.0		|| This is a sorting class(cont)||
||	designed for Windows 10. It is intended to sort multiple objects of the	||
||	same type. Multiple sorting algorithms are planned for implementation.	||
||																			||
||	04/26/2019	||	06:55PM	||	0.2.0.0		|| two functions have been		||
||	implemented, but yet untested. Quicksort Algorithm is implemented.		||
||																			||
||	04/27/2019	||	06:55PM	||	0.2.1.2		|| quicksort algorithm is fixed	||
||	and selection sort is implemented.										||
||																			||
||	Version History:														||
||	Date:		||	Time:	||	Version:	||	Description:				||
||	04/26/2019	||	05:37PM	||	0.8.8.8		|| 8 functions have been(cont)	||
||	implemented and fixed in this version.									||
||																			||
----------------------------------------------------------------------------*/

#pragma once
#ifndef ABSORTINGCLASS_H
#define ABSORTINGCLASS_H
#include <vector>
#include <iostream>//Temporary
#include <ctime>//for array randomization
#include <iomanip>
#include <algorithm>
using namespace std;

namespace AB_SortingClass
{
	const int MAX_SIZE = 1000;
	//Array
	void quicksort(int* list, int low, int high);
	int partition(int* list, int low, int high);
	void selectionSort(int* list,int low, int size);
	void optQuicksort(int* list, int low, int high, int ssSize);
	int optPartition(int* list, int low, int high);
	void optSelectionSort(int* list, int low, int size);


	//randomizes an array.
	void randomizeArr(int* list, int size);
	void printArr(int* arr, int size);
}
#endif // !SORTINGCLASS_H
//--------------------------------End of File---------------------------------

SourceCode/ABSortingClass.cpp
/*--------------------------------------------------------------------------||
||	Programmer:			Alexander Burton									||
||	Date Created:		4/26/2019											||
||	Version:			0.8.8.8												||
||	File:				ABSortingClass.cpp									||
||	Compiler Info:															||
||		Hardware:															||
||			MB:																||
||				Manu:		BaseBoard Manufacturer	ASUSTeK COMPUTER INC.	||
||				Chipset:	BaseBoard Product	P8Z77-V PRO					||
||																			||
||			CPU:															||
||				Chipset:	Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz			||
||				Clock Rate:	3401 Mhz										||
||				Cores:		4												||
||				processors:	8												||
||																			||
||			RAM:															||
||				Manu:				Kingston								||
||				Capacity(indiv):	4GB										||
||				Count:				4										||
||				Capacity(total):	16.0GB									||
||				Type:				240-Pin DDR3 SDRAM DDR3 1333 (PC3 10600)||
||																			||
||		Software:															||
||				OS:			Windows 10 Home Edition							||
||				IDE:		Visual Studio 2019								||
||																			||
||	Functions:																||
||																			||
||		void quicksort(int* list, int low, int high);						||
||		int partition(int* list, int low, int high);						||
||		void selectionSort(int* list,int low, int size);					||
||		void optQuicksort(int* list, int low, int high, int ssSize);		||
||		int optPartition(int* list, int low, int high);						||
||		void optSelectionSort(int* list, int low, int size);				||
||																			||
||		void randomizeArr(int* list, int size);								||
||		void printArr(int* arr, int size);									||
||																			||
||	Variables:																||
||		N/A																	||
||	Notes:																	||
||	Version History:														||
||	Date:		||	Time:	||	Version:	||	Description:				||
||	04/26/2019	||	05:37PM	||	0.0.0.0		|| This is a sorting class(cont)||
||	designed for Windows 10. It is intended to sort multiple objects of the	||
||	same type. Multiple sorting algorithms are planned for implementation.	||
||																			||
||	04/26/2019	||	06:55PM	||	0.2.0.0		|| two functions have been		||
||	implemented, but yet untested. Quicksort Algorithm is implemented.		||
||																			||
||	04/27/2019	||	06:55PM	||	0.2.1.2		|| quicksort algorithm is fixed	||
||	and selection sort is implemented.										||
||																			||
||	Version History:														||
||	Date:		||	Time:	||	Version:	||	Description:				||
||	04/26/2019	||	05:37PM	||	0.8.8.8		|| 8 functions have been(cont)	||
||	implemented and fixed in this version.									||
||																			||
----------------------------------------------------------------------------*/
#include "ABSortingClass.h"

namespace AB_SortingClass
{

	/*----------------------------------------------------------------------||
	||	Programmer:			Alexander Burton								||
	||	Date Created:		4/26/2019										||
	||	Version:			0.1.1.1											||
	||	Function: Quicksort													||
	||																		||
	||	Function calls:														||
	||		void partition(T* list[], T* low, T* high)						||
	||		quicksort(list, low, pivot - 1)									||
	||																		||
	||	Called by:															||
	||		int main()														||
	||																		||
	||	Notes:																||
	||		When calling the function, subtract 1 from the third parameter.	||
	||																		||
	||	Version History:													||
	||	Date:		||	Time:	||	Version:	||	Description:			||
	||	04/26/2019	||	05:37PM	||	0.1.1.0		|| This is a quicksort(cont)||
	||	algorithm that partitions the data set, which entails moving all	||
	||	elements greater than a given pivot to one side, and elements less	||
	||	than to the other until the sort size is zero.						||
	||	n log(n) best case time complexity.									||
	||																		||
	||	04/28/2019	||	06:33PM	||	0.1.1.1		|| Fixed an issue with(cont)||
	||	sorting.															||
	||																		||
	||----------------------------------------------------------------------*/
	void quicksort(int* list, int low, int high)
	{
		if (low < high)
		{
			int pivot = partition(list, low, high);
			quicksort(list, low, pivot - 1);
			quicksort(list, pivot + 1, high);
		}
	}
	/*----------------------------------------------------------------------||
	||	Programmer:			Alexander Burton								||
	||	Date Created:		4/26/2019										||
	||	Version:			0.1.1.1											||
	||	Function: Partition													||
	||																		||
	||	Function calls:														||
	||																		||
	||	Called by:															||
	||		void quicksort(int* list, int low, int high)					||
	||	Notes:																||
	||	Version History:													||
	||	Date:		||	Time:	||	Version:	||	Description:			||
	||	04/26/2019	||	05:37PM	||	0.0.1.0		|| This is a partition(cont)||
	||	algorithm that partitions the data set, which entails moving all	||
	||	elements greater than a given pivot to one side, and elements less	||
	||	than to the other until the sort size is zero.						||
	||	n log(n) best case time complexity.									||
	||																		||
	||	04/28/2019	||	06:33PM	||	0.1.1.1		|| Fixed an issue with(cont)||
	||	sorting.															||
	||																		||
	||----------------------------------------------------------------------*/
	int partition(int* list, int low, int high)
	{
		int pivot = list[high];
		int i = low;
		for (int j = low; j < high; j++)
		{
			if (list[j] < pivot)
			{
				swap(list[i], list[j]);
				i++;
			}
		}
		swap(list[i], list[high]);
		return i;
	}
	/*----------------------------------------------------------------------||
	||	Programmer:			Alexander Burton								||
	||	Date Created:		4/26/2019										||
	||	Version:			0.1.1.0											||
	||	Function: Partition													||
	||																		||
	||	Function calls:														||
	||																		||
	||	Called by:															||
	||																		||
	||	Notes:																||
	||	Version History:													||
	||	Date:		||	Time:	||	Version:	||	Description:			||
	||	04/26/2019	||	05:37PM	||	0.0.1.0		|| This is a sorting(cont)	||
	||	algorithm that goes through 2 loops(nested) comparing two values	||
	||	and swapping relevent elements until the entire array is sorted.	||
	||	Exponential time.(n^2)												||
	||																		||
	||----------------------------------------------------------------------*/
	void selectionSort(int* list, int low, int size)
	{
		bool swapped = false;
		for (int i = low; i < size; i++)
		{
			swapped = false;
			for (int j = i; j < size; j++)
			{
				if (list[i] > list[j])
				{
					swap(list[i], list[j]);
					swapped = true;
				}
			}
			if (swapped == false)
				break;
		}
	}

	/*----------------------------------------------------------------------||
	||	Programmer:			Alexander Burton								||
	||	Date Created:		4/26/2019										||
	||	Version:			0.1.1.1											||
	||	Function: optimized Quicksort										||
	||																		||
	||	Function calls:														||
	||		void optPartition(list[], low, high)							||
	||		void optQuicksort(list, low, pivot)								||
	||		void optSelectionSort(int* list, int low, int size)				||
	||																		||
	||	Called by:															||
	||																		||
	||	Notes:																||
	||		When calling the function, subtract 1 from the third parameter.	||
	||																		||
	||	Version History:													||
	||	Date:		||	Time:	||	Version:	||	Description:			||
	||	04/26/2019	||	05:37PM	||	0.1.1.0		|| This is a quicksort(cont)||
	||	algorithm that partitions the data set, which entails moving all	||
	||	elements greater than a given pivot to one side, and elements less	||
	||	than to the other until the sort size is zero.						||
	||	n log(n) best case time complexity.									||
	||																		||
	||	04/28/2019	||	06:33PM	||	0.1.1.1		|| Fixed an issue with(cont)||
	||	sorting.															||
	||																		||
	||----------------------------------------------------------------------*/
	void optQuicksort(int* list, int low, int high, int ssSize)
	{
		if (low < high)//determine if left and right indicies overlap.
		{
			//if the distance between indices are greater than ssSize, then
			//Perform usual quicksort divide and conquer.
			if ((high - low) > ssSize)
			{
				//get pivot from partition.
				int pivot = optPartition(list, low, high);
				//quicksort left and right sides of array.
				optQuicksort(list, low, pivot - 1, ssSize);
				optQuicksort(list, pivot + 1, high, ssSize);
			}
			else
			{
				//distance between indicies is less than or equal to ssSize.
				//perform selection sort.
				optSelectionSort(list, low, high + 1);
			}
		}
	}
	/*----------------------------------------------------------------------||
	||	Programmer:			Alexander Burton								||
	||	Date Created:		4/26/2019										||
	||	Version:			0.1.1.1											||
	||	Function: optPartition												||
	||																		||
	||	Function calls:														||
	||																		||
	||	Called by:															||
	||		void optQuicksort(int* list, int low, int high)					||
	||	Notes:																||
	||	Version History:													||
	||	Date:		||	Time:	||	Version:	||	Description:			||
	||	04/26/2019	||	05:37PM	||	0.0.1.0		|| This is a partition(cont)||
	||	algorithm that partitions the data set, which entails moving all	||
	||	elements greater than a given pivot to one side, and elements less	||
	||	than to the other until the sort size is zero.						||
	||	n log(n) best case time complexity.									||
	||																		||
	||	04/28/2019	||	06:33PM	||	0.1.1.1		|| Fixed an issue with(cont)||
	||	sorting.															||
	||																		||
	||----------------------------------------------------------------------*/
	int optPartition(int* list, int low, int high)
	{
		//set the pivot to high.
		int pivot = list[high];
		//starting position.
		int i = low;
		for (int j = low; j < high; j++)
		{
			if (list[j] < pivot)
			{
				swap(list[i], list[j]);
				i++;
			}
		}
		swap(list[i], list[high]);
		return i;
	}
	/*----------------------------------------------------------------------||
	||	Programmer:			Alexander Burton								||
	||	Date Created:		4/26/2019										||
	||	Version:			0.1.1.0											||
	||	Function: Partition													||
	||																		||
	||	Function calls:														||
	||																		||
	||	Called by:															||
	||		void optQuicksort(int* list, int low, int high)					||
	||																		||
	||	Notes:																||
	||	Version History:													||
	||	Date:		||	Time:	||	Version:	||	Description:			||
	||	04/26/2019	||	05:37PM	||	0.0.1.0		|| This is a sorting(cont)	||
	||	algorithm that goes through 2 loops(nested) comparing two values	||
	||	and swapping relevent elements until the entire array is sorted.	||
	||	Exponential time.(n^2) for use with optQuicksort					||
	||																		||
	||----------------------------------------------------------------------*/
	void optSelectionSort(int* list, int low, int size)
	{
		bool swaps;
		for (int i = low; i < size; i++)
		{
			swaps = false;
			for (int j = i; j < size; j++)
			{
				if (list[i] > list[j])
				{
					swaps = true;
					swap(list[i], list[j]);
				}
			}
			if (swaps == false)
				break;
		}
	}
	/*----------------------------------------------------------------------||
	||	Programmer:			Alexander Burton								||
	||	Date Created:		4/26/2019										||
	||	Version:			0.1.1.0											||
	||	Function: Partition													||
	||																		||
	||	Function calls:														||
	||																		||
	||	Called by:															||
	||			void randomizeArr(int* list, int size)						||
	||																		||
	||	Notes:																||
	||	Version History:													||
	||	Date:		||	Time:	||	Version:	||	Description:			||
	||	04/26/2019	||	05:37PM	||	0.1.1.0		|| This is a randomizer		||
	||	fucntion using a left index and a right index that point to an		||
	||	that has been passed in via pointer.								||
	||																		||
	||----------------------------------------------------------------------*/
	void randomizeArr(int* list, int size)
	{
		//Left index is the first index array. Right is the second.
		int leftIndex = 0, rightIndex = 0;
		//randomize array and vector
		for (int i = 0; i < size; ++i)
		{
			srand((int)time(0) + i);//randomize seed
			leftIndex = rand() % size;//left index is randomized
			for (int j = 0; j < size; ++j)
			{
				srand((int)time(0) + i + j);//randomize seed
				rightIndex = rand() % size;//right index is randomized
				while (rightIndex == leftIndex)
				{
					rightIndex = rand() % size;//right index is randomized
				}
				//swap array elements
				int temp = list[leftIndex];
				list[leftIndex] = list[rightIndex];
				list[rightIndex] = temp;
			}
		}
	}
	void printArr(int* arr, int size)
	{
		for (int i = 0; i < size; i++)
		{
			cout << " * " << setw(5) << arr[i];
			if ((i + 1) % 10 == 0)
				cout << " *" << endl;
		}
	}
};
//--------------------------------End of File---------------------------------